pinot/regressors/biophysical_regressor.py

```diff
@@ -0,0 +1,78 @@
+# =============================================================================
+# IMPORTS
+# =============================================================================
+import torch
+import pinot
+import abc
+import math
+import numpy as np
+
+class BiophysicalRegressor(torch.nn.Module):
+    r""" Biophysically inspired model
+
+    Parameters
+    ----------
+
+    log_sigma : `float`
+        ..math:: \log\sigma observation noise
+
+    base_regressor : a regressor object that generates a latent F
+
+
+    """
+
+    def __init__(self, base_regressor=None):
+        super(BiophysicalRegressor, self).__init__()
+        self.base_regressor = base_regressor
+        self.log_sigma_measurement = torch.nn.Parameter(torch.zeros(1))
+
+
+	def g(self, func_value=None, test_ligand_concentration=1e-3):
+	    return 1 / (1 + torch.exp(-func_value) / test_ligand_concentration)
+
+
+	def condition(self, h=None, test_ligand_concentration=1e-3, **kwargs):
+		distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+		#we sample from the latent f to push things through the likelihood
+		#Note: if we do this, in order to get good estimates of LLK we may need to draw multiple samples
+		f_sample = distribution_base_regressor.rsample()
+		mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+		sigma_m = torch.exp(self.log_sigma_measurement)
+		distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+		return distribution_measurement
+
+
+	def loss(self, h=None, y=None, test_ligand_concentration=1e-3, **kwargs):
+		distribution_measurement = self.condition(h=h, test_ligand_concentration=test_ligand_concentration, **kwargs)
+		loss_measurement = -distribution_measurement.log_prob(y)
+		return loss_measurement
+
+
+   	def marginal_sample(self, h=None, n_samples=100, test_ligand_concentration=1e-3, **kwargs):
+   		distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+   		samples_measurement = []
+   		for ns in range(n_samples):
+   			f_sample = distribution_base_regressor.rsample()
+	        mu_m = self.g(func_value=f_sample, test_ligand_concentration=test_ligand_concentration)
+	        sigma_m = torch.exp(self.log_sigma_measurement)
+	        distribution_measurement = torch.distributions.normal.Normal(loc=mu_m, scale=sigma_m)
+    		samples_measurement.append(distribution_measurement.sample())
+    	return samples_measurement
+
+
+    def marginal_loss(self, h=None, y=None, test_ligand_concentration=1e-3, n_samples=10, **kwargs):
+    	"""
+    	sample n_samples often from loss in order to get a better approximation
+    	"""
+    	distribution_base_regressor = self.base_regressor.condition(h, **kwargs)
+    	marginal_loss_measurement = 0
+    	for ns in range(n_samples):
+    		for ns in range(n_samples):
```

Comment:
this seems strange to me. perhaps it could have two different params, one for the inner vs. outer loop?

Reply:
elaborate please

Reply:
one is for sampling from the base_regressor posterior (inner loop), and the other is for sampling from the loss function (outer loop). in this case, we've given them always the same number of samples for each, but it doesn't have to be that way, right?

actually I'm really confused now because I see that you're just setting f_sample, so the inner loop isn't doing anything; you're just using the results from the most recent loop.

maybe I'm misunderstanding what this is doing.

Reply:
similar question applies for the for loop inside `marginal_sample` actually (lines 54-55)

Reply:
I am sampling both each time... I don;t see the need for an inner and an outer loop I just go:
```
dG ~ N(mu_regressor, sigma_regressor)
m ~ N(g(dG,c), sigma_measurement)
```

and voila, that is my generative model.

Reply:
If you look closely, for the marginal loss I am actually trying to marginalize out the deltaG, that might correspond to your inner loop idea.

Reply:
that makes more sense to me, but there are nested loops inside that function, not a single loop

Reply:
I'm just asking what the function of these lines are, found in both marginal_loss and marginal_sample
```
   		for ns in range(n_samples):
   			f_sample = distribution_base_regressor.rsample()
```

Reply:
> that makes more sense to me, but there are nested loops inside this function (as currently written), not a single loop

Yeah, that's what's known as a copy-paste bug. I commited a change, good catch.